Entities/SLTerrain.cpp

```diff
@@ -1,1436 +1,474 @@
-
-            if (testPixel != g_MaskColor)
-            {
-                // Only add PixelMO if we're not due to skip any
-                if (makeMOPs && matPixel != g_MaterialAir && colorPixel != g_MaskColor && ++skipCount > skipMOP && MOPDeque.size() < maxMOPs)
-                {
-                    skipCount = 0;
-                    sceneMat = g_SceneMan.GetMaterialFromID(matPixel);
-                    spawnMat = sceneMat->GetSpawnMaterial() ? g_SceneMan.GetMaterialFromID(sceneMat->GetSpawnMaterial()) : sceneMat;
-                    // Create the MOPixel based off the Terrain data.
-                    pPixel = new MOPixel(colorPixel,
-                                         spawnMat->GetPixelDensity(),
-                                         Vector(terrX, terrY),
-                                         Vector(),
-                                         new Atom(Vector(), spawnMat->GetIndex(), 0, colorPixel, 2),
-                                         0);
-
-                    pPixel->SetToHitMOs(false);
-                    MOPDeque.push_back(pPixel);
-                    pPixel = 0;
-                }
-
-                // Clear the terrain pixels
-                if (matPixel != g_MaterialAir)
-                    putpixel(m_pMainBitmap, terrX, terrY, g_MaterialAir);
-				if (colorPixel != g_MaskColor)
-				{
-					putpixel(m_pFGColor->GetBitmap(), terrX, terrY, g_MaskColor);
-					g_SceneMan.RegisterTerrainChange(terrX, terrY, 1, 1, g_MaskColor, false);
+	int SLTerrain::Save(Writer &writer) const {
+		SceneLayer::Save(writer);
+
+		// Only write the background texture info if the background itself is not saved out as a file already.
+		if (m_BGColorLayer->IsLoadedFromDisk()) {
+			writer.NewPropertyWithValue("BGColorLayer", m_BGColorLayer.get());
+		} else {
+			writer.NewPropertyWithValue("BackgroundTexture", m_DefaultBGTextureFile);
+		}
+
+		// Only write the procedural params if the foreground itself is not saved out as a file already.
+		if (m_FGColorLayer->IsLoadedFromDisk()) {
+			writer.NewPropertyWithValue("FGColorLayer", m_FGColorLayer.get());
+		} else {
+			// Layer data is not saved into a bitmap file yet, so just write out the procedural params to build the terrain.
+			for (const TerrainFrosting *terrainFrosting : m_TerrainFrostings) {
+				writer.NewPropertyWithValue("AddTerrainFrosting", terrainFrosting);
+			}
+			for (const TerrainDebris *terrainDebris : m_TerrainDebris) {
+				writer.NewPropertyWithValue("AddTerrainDebris", terrainDebris);
+			}
+			for (const TerrainObject *terrainObject : m_TerrainObjects) {
+				// Write out only what is needed to place a copy of this in the Terrain
+				writer.NewProperty("PlaceTerrainObject");
+				writer.ObjectStart(terrainObject->GetClassName());
+				writer.NewPropertyWithValue("CopyOf", terrainObject->GetModuleAndPresetName());
+				writer.NewPropertyWithValue("Position", terrainObject->GetPos());
```

Comment:
Why would I want/need to make modifiable CopyOfs here?

Reply:
Cause it can be nice for ini users to be able to do CopyOf and modify things instead of the hodgepodge of functionality we have right now where you can do that for some stuff but not for others?

Not a big deal either way though.